akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala

```diff
@@ -0,0 +1,140 @@
+
+  "A FlowDeathPact" must {
+
+    "complete when contrActor terminates" in {
+      val devil = system.actorOf(TestActors.echoActorProps)
+
+      val deathPact: FlowDeathPact[String] = FlowDeathPact.completeStage(devil)
+      val probe = Source.maybe[String].via(deathPact).runWith(TestSink.probe)
+
+      probe.expectSubscription()
+      probe.expectNoMsg(100.millis)
+
+      devil ! PoisonPill
+      probe.expectComplete()
+    }
+
+    "fail with DeathPactException when contrActor terminates" in {
+      val devil = system.actorOf(TestActors.echoActorProps)
+
+      val deathPact: FlowDeathPact[String] = FlowDeathPact.failStage(devil)
+      val probe = Source.maybe[String].via(deathPact).runWith(TestSink.probe)
+
+      probe.expectSubscription()
+
+      devil ! PoisonPill
+      val ex = probe.expectError()
+      ex.getClass should ===(classOf[DeathPactException])
+    }
+
+    "inform contrActor that FlowDeathPact was Signed" in {
+      val contrActor = TestProbe()
+
+      val deathPact: FlowDeathPact[String] = FlowDeathPact.failStage(contrActor.ref)
+      val probe = Source.empty.via(deathPact).runWith(TestSink.probe)
+
+      contrActor.expectMsg(FlowDeathPact.Signed)
+
+      probe.expectSubscriptionAndComplete()
+    }
+
+    "inform contrActor that when stream Failed" in {
+      val contrActor = TestProbe()
+
+      val deathPact: FlowDeathPact[String] = FlowDeathPact.failStage(contrActor.ref)
+      val cause = new Exception("BOOM!")
+      val probe = Source.failed(cause).via(deathPact).runWith(TestSink.probe)
+
+      contrActor.expectMsg(FlowDeathPact.Signed)
+      val failed = contrActor.expectMsgType[FlowDeathPact.Failed]
+      failed.cause should ===(cause)
+
+      probe.expectSubscription().request(1)
+      probe.expectError(cause)
+    }
+
+    "inform contrActor that when stream Completed" in {
+      val contrActor = TestProbe()
+
+      val deathPact: FlowDeathPact[String] = FlowDeathPact.failStage(contrActor.ref)
+      val probe = Source.empty.via(deathPact).runWith(TestSink.probe)
+
+      contrActor.expectMsg(FlowDeathPact.Signed)
+      contrActor.expectMsg(FlowDeathPact.Completed)
+      info("Stage sender: " + contrActor.lastSender)
+      probe.expectSubscriptionAndComplete()
+    }
+
+  }
+
+}
+
+/** INTERNAL API */
+/*private[akka]*/ final class FlowDeathPact[T](ref: ActorRef, onTerminated: GraphStageLogic ⇒ Unit)
+  extends GraphStage[FlowShape[T, T]] {
+
+  val in = Inlet[T]("FlowDeathPact.in")
+  val out = Outlet[T]("FlowDeathPact.out")
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) 
+    with InHandler with OutHandler {
+
+    override def preStart(): Unit = {
+      getStageActor({
+        case (_, Terminated(`ref`)) ⇒ onTerminated(this)
+        case _                      ⇒ // ignore...
+      })
+      stageActor.watch(ref)
+      ref.tell(FlowDeathPact.Signed, stageActor.ref)
```

Comment:
For that, the ActorRef is enough. Plus, the built-in watch mechanism is much more reliable and works over the wire. Hence I think it is much better to expose the stage ActorRef and let the other side watch it if it wants to.